xarray/core/duck_array_ops.py

```diff
@@ -261,7 +261,7 @@ def where_method(data, cond, other=dtypes.NA):
 
 
 def fillna(data, other):
-    return where(isnull(data), other, data)
+    return where(notnull(data), data, other)
 
```

Comment:
Is there a reason why you had to flip this?

Reply:
Is this worth fixing in pint? It seems like it would make sense to determine units in `where` a symmetrical way, similar to operations like `+`.

(We can keep this change for now -- I don't see much downside -- but please add a comment documenting why, so we will know when or if this could be changed in the future)

Reply:
actually, the current implementation of `where` is consistent with the behaviour of `+`:
```python
In [1]: import pint 
   ...: import numpy as np 
   ...: ureg = pint.UnitRegistry() 
   ...:  
   ...: a = [12, 10, 9] * ureg.m 
   ...: b = [15, 9, 11] * ureg.mm 
   ...: a + b, b + a
Out[1]: 
(array([12.015, 10.009,  9.011]) <Unit('meter')>,
 array([12015., 10009.,  9011.]) <Unit('millimeter')>)
```

comment: done

Reply:
Pint's implementation of `where` indeed behaves like `+` by using the units of the first argument:

```python
a = [[1, 2], [3, 4]] * ureg.m
b = [[-100, -200], [-300, -400]] * ureg.cm
m = np.array([[True, False], [False, True]])
print(a + b)
print(b + a)
print(np.where(m, a, b))
print(np.where(m, b, a))
```
```
[[0.0 0.0] [0.0 0.0]] meter
[[0.0 0.0] [0.0 0.0]] centimeter
[[1.0 -2.0] [-3.0 4.0]] meter
[[-100.0 200.0] [300.0 -400.0]] centimeter
```

(this only really matters though if the `identical` check is particular about exact unit as discussed below)

Reply:
I'm a little surprised that arithmetic is not commutative in pint, but at least it's internally consistent! I suspect this is likely to cause issues in other places where we use duck typing.

I guess this is a bigger issue, but I think it would be worth considering adding some sort of unit casting hierarchy into pint's unit registries. It almost doesn't matter what exactly the rules are as long as it exists.

For a library like xarray that rely upon duck typing, it is nice to be able to rely upon properties like arithmetic being commutative to reduce complexity. When really needed we can add special case logic, but it means untested code paths are likely to break in unexpected ways.

Reply:
@shoyer That's a good point! Having
```python
(a + b).units == (b + a).units
```
(and likewise for other commutative bin-ops) would be a good invariant. I'll raise an issue on Pint.